### ISSUE TITLE
IBM z/Architecture Principles of Operation DOC bug

### ISSUE THREAD

**Issue by Fish-Git** [SDL-Hercules-390/hyperion#299]
IBM manual SA22-7832-12 "z/Architecture Principles of Operation" contains a documentation bug.

On page 20-24 it shows the format for the CONVERT FROM FIXED instructions CDFTR (B951) and CXFTR (B959) as _apparently_ being format RRF-e:
![docbug-current](https://user-images.githubusercontent.com/1486786/81138602-6d5c3280-8f17-11ea-9484-09dbf83251bf.jpg)
However, the two mentioned instructions (B951=CDFTR, B959=CXFTR), according to Figure B-3 ("Instructions Arranged by Opcode  (Part 11 of 24)") on Appendix B page B-63 are listed as _actually_ being format _**RRE**_, not RRF-e.

So page 20-24 _**should**_ actually look like the following:
![docbug-SHOULDBE](https://user-images.githubusercontent.com/1486786/81138800-1e62cd00-8f18-11ea-9bac-a2dd60b9c736.jpg)
This bug _should_ be reported to IBM, but I don't know how to do that.

This GitHub Issue was created simply as an FYI ("For Your Information") and to ask for someone's help regarding going about reporting this doc bug to IBM.

Thanks!

**Comment by Fish-Git** [comment]
FYI to Harold Grovesteen (@s390guy):

Looking at the `SATK-0\asma\msl\` **`all-insn.msl`**  and **`s390x-insn.msl`** files, it looks like your ASMA product contains contains the same bug that Hercules had (that was just recently fixed by commit 0f382a7875b85359cb6881beeacf26e1324ee562).

**Comment by mcisho** [comment]
Perhaps the instruction description is correct, and Figure B-3 is wrong? The following is from a z/OS 2.2 assembly:-

000000 0000 0000                      2          CDFTR 4,6
** ASMA175S Delimiter error, expected comma -
** ASMA435I Record ...
000004 0000 0000                      3          CXFTR 4,6
** ASMA175S Delimiter error, expected comma -
** ASMA435I Record ...
000008 B951 6A48                      4          CDFTR 4,6,8,10
00000C B959 6A48                      5          CXFTR 4,6,8,10

Appears that the assembler agrees with the instruction description.
Ian

**Comment by Fish-Git** [comment]
> Perhaps the instruction description is correct, and Figure B-3 is wrong?

Then _ALL_ of the appendices are wrong. Appendix B1, B2 _and_ B3 _all_ list the instructions as being RRE format, _not_ RRF-e.

And I just noticed another bug (discrepancy) too: while Figure 20-9 "Summary of Decimal Instructions" on page 20-17 correctly lists instruction CXFTR (B959) as being format RRE, it _incorrectly_ lists instruction CDFTR (B951) as being format RRF-e:

![docbug-FIGURE-20-9](https://user-images.githubusercontent.com/1486786/81232057-0e46fe00-8fa9-11ea-8b29-f1680ab42154.jpg)

So it's obvious the manual is wrong in more than one place. It's obvious the manual _**does**_ contain doc bugs. The only question is _where (i.e. who is right and who is wrong)._

IMHO, based on the totality of evidence (which I believe is in my favor), both instructions are actually format RRE instructions, _not_ RRF-e. As far as why the assembler kicks out an error when format RRF-e isn't used, I suspect the reason is it's simply wrong too.(*) IBM certainly isn't immune from making mistakes after all.

_**Bottom line:** &nbsp;Something_ is wrong somewhere, so either way, it should be reported to IBM.

Does anyone know how to do that?
___
(*) Have you tried the same test for the CXGTR/CXGTRA (B3F9) and CDGTR/CDGTRA (B3F1) instructions? Their format depends on which opcode is used. The CXGTR/CDGTR mnemonics are format RRE whereas the CXGTRA/CDGTRA mnemonics are format RRF-e. Each are the same exact instruction, but use a different format depending on how they're coded in assembler.

Such is NOT the case _(or shouldn't be!)_ for the CDFTR/CXFTR instructions however. They should _both_ be format RRE.

In my opinion. &nbsp; `;-)`

**Comment by mcisho** [comment]
All six of the CONVERT FROM FIXED instructions assemble correctly if they are coded as described in the instruction description, i.e. two operands for GDGTR and CXGTR (RRE), and four operands for CDGTRA, CXGTRA, CDFTR and CXFTR (RRF-e).

I would expect Figures B-1, B-2 and B-3 to match, they are after all the same set of information sorted into different orders.

There is definitely a documentation bug, but IMHO it is Appendix B that is in error, and the instruction description and the assembler that are correct. However, until IBM pronounces, or someone produces a test program that shows CDFTR/CXFTR are RRE or RRF-e,  we all remain in the dark.

The final page of the PoP is titled 'Communicating Your Comments to IBM', and provides an email address for comments on errors, etc.

**Comment by Fish-Git** [comment]
> There is definitely a documentation bug, ...

Good! At least we agree on something. &nbsp; `:)`

> ... but IMHO it is Appendix B that is in error, and the instruction description and the assembler that are correct.

And that's where we disagree.

I mean, look at the mnemonics and the description of the instructions themselves! The CDGTR/CXGTR mnemonics are for the 64-bit variation of the instruction (hence the 'G' in their name) and the CDFTR/CXFTR mnemonics are for the 32-bit variation (hence the 'F' in their name). That's standard IBM mnemonic naming as mentioned on pages pages, 1-3, 1-5, 1-6, as well as on page 7-7 too.

Thus if CDGTR/CXGTR are RRE format and _only_ support two operands (but not four), then it only makes sense that CDFTR/CXFTR would be the same way, and that it is the trailing 'A' on the CDGTRA/CXGTRA mnemonics that indicates the instruction should instead be assembled in RRF-e format, thus requiring fours operands instead of two.(*)

I think both you and IBM's assembler are wrong Ian, and both Appendix B and I are right.

> The final page of the PoP is titled 'Communicating Your Comments to IBM', and provides an email address for comments on errors, etc.

_(Doh!)_ Thanks. I completely forgot about that.

I'll report it to IBM right away and will let you know what they say (if they even bother to respond!).
___
(*) That's the only difference between the two after all (CDGTR vs. CDGTRA and CXGTR vs. CXGTRA). Both of their opcodes are the exactly the same. The only difference is their assembler format. Hence the difference mnemonic.

**Comment by mcisho** [comment]
If you're correct you also need to change the CONVERT TO FIXED instructions CFDTR and CFXTR.

However, before you do please refer to Figure 20-9 Summary of Decimal Instruction (which is a subset of Appendix B). Even more confusion! Instructions the are RRE are shown as RRF-e (as they also are in Appendix B).

Personally, I don't think you should make changes in this area until the confusion is resolved.

**Comment by Fish-Git** [comment]
> However, before you do please refer to Figure 20-9 Summary of Decimal Instruction (which is a subset of Appendix B). Even more confusion! Instructions the are RRE are shown as RRF-e (as they also are in Appendix B).

Isn't that what I **[already mentioned?](https://github.com/SDL-Hercules-390/hyperion/issues/299#issuecomment-624923732)** &nbsp; `:)`

**Comment by Fish-Git** [comment]
More confusion!

* **["z/Architecture Reference Summary" (SA22-7871-10)](https://www.ibm.com/support/pages/zarchitecture-reference-summary-0)**

While it lists both CDFTR and CXFTR as being **RRE** format (_not_ RRF-e), it also lists "**R1,M3,R2,M4**" _(RRF-e format!)_ in the (assembler) **Operands** column!
___
![docbug-ref-8](https://user-images.githubusercontent.com/1486786/81310819-7513f800-9039-11ea-89a8-c4fec03cfaf6.jpg)
___
![docbug-ref-12](https://user-images.githubusercontent.com/1486786/81310838-7ba26f80-9039-11ea-8548-c12c0ccb1db3.jpg)
___
This _**definitely**_ needs to be reported!

_(But I still say the evidence is (slightly!) in my favor)_ &nbsp; `;-)`

**Comment by Fish-Git** [comment]
> Personally, I don't think you should not make changes in this area until the confusion is resolved.

In the interest of erring on the side of caution, I am going to have to agree with this and revert **[my recent commit](0f382a7875b85359cb6881beeacf26e1324ee562)** until I can get a definitive answer from IBM.

**Comment by mcisho** [comment]
The CDGTR and CXGTR instructions first appear in zPop-05 (April 2007) with the introduction of the Decimal-Floating-Point Facility. The CDGTRA, CXGTRA, CDFTR and CXFTR instructions first appear in zPop-08 (August 2010) with the introduction of the Floating-Point Extension Facility; the CDFTR and CXFTR instruction format confusion dates from their appearance.

The CDGTRA, CXGTRA, CDFTR and CXFTR instructions have always been documented as having four operands, irrespective of their documented format.

Presumably the CxGTRA instructions were introduced to distinguish the use of the new facility from the existing CxGTR instructions. However, as there were no existing CxFTR instructions, the CxFTR instructions have always used the new facility.

A **CxGTRA r1,r2,0,0** instruction is functionally identical to a **CxGTR r1,r2** instruction. The functionality of the CxGTRA instructions only change when the third and fourth operands are non-zero.

**Comment by Fish-Git** [comment]
Yes, I am now beginning to lean in your direction. I am beginning to believe you and IBM's assembler are right and I and Appendix B et al. are wrong, based on the very same logic: the two instructions in question were introduced with the Floating-Point Extension Facility, whereas CDGTR and CXGTR existed prior to that. So it's beginning to make sense now that CDGTR and CXGTR would be RRE format whereas _ALL_ of the new instructions introduced with the Floating-Point Extension Facility (which includes the two in question: CDFTR and CXFTR) are all RRF-e format. That makes more sense than anything.

Confusing? Absolutely!

And _regardless_ of who is right and who is wrong (but as I said I'm now beginning to believe you are right and I am wrong), the IBM manuals in question are _**definitely wrong!**_ That is to say, a DOC BUG _does_ definitely exist in the two manuals. And it's confusing.

So I am going to go ahead and report this.

I'll let you know what IBM says if they bother to respond.

Thanks for your thoughtful input, Ian, and especially for your patience with my occasional obstinance regarding such matters. &nbsp; `:)`

**Comment by Fish-Git** [comment]
Wow. That didn't take long!

I sent an email to IBM yesterday and got their response _today!_

And the results are ...................... (drum roll, please) ...........................

............ _**both instructions are indeed RRF-e format, just as you said!**_

So congratulations, Ian! You were right and I was wrong! Mystery solved!

Their full response was:

> David,
> 
> Instructions CDFTR and CXFTR are provided with the floating-point extension facility.
> Both instructions include M3 and M4 fields. Therefore, the instruction format is RRF-e
> (see "Instruction Formats" in chapter 5 of SA22-7832-12).
> 
> I have added the mistakes you've identified to the list of corrections
> to be made in the next version of each document. 
> 
> Thank you for your feedback,
> 
> Bruce Giamei

I am going to close this issue now. Apologies for the false alarm, but hey, we at least _did_ do _some_ good. We _did_ identify a confusing IBM Documentation bug, and they did promise to make the needed corrections in the next version of each document, so _some_ definite good did come out of all this.

_**THANK YOU**, Ian,_ for sticking it out with me and for your many helpful/thoughtful comments.

_MUCH_ appreciated, my friend! &nbsp; `:)`

**Comment by Fish-Git** [comment]
> FYI to Harold Grovesteen (@s390guy):
> 
> Looking at the `SATK-0\asma\msl\` **`all-insn.msl`** and **`s390x-insn.msl`** files, it looks like your ASMA product contains the same bug that Hercules had (that was just recently fixed by commit [0f382a7](https://github.com/SDL-Hercules-390/hyperion/commit/0f382a7875b85359cb6881beeacf26e1324ee562)).

FYI to Harold Grovesteen (@s390guy): &nbsp; _**I WAS WRONG!**_

Your ASMA product does _not_ need to be changed and is in fact _100% correct, as-is._

My apologies for the false alarm.
